File: myliquor-backend/src/main/java/memo/liquor/myliquor/service/LiquorService.java
package memo.liquor.myliquor.service;

import lombok.RequiredArgsConstructor;
import memo.liquor.myliquor.controller.CategoryResponse;
import memo.liquor.myliquor.controller.LiquorResponse;
import memo.liquor.myliquor.controller.SaveLiquorRequest;
import memo.liquor.myliquor.entity.Category;
import memo.liquor.myliquor.entity.Liquor;
import memo.liquor.myliquor.repository.CategoryRepository;
import memo.liquor.myliquor.repository.LiquorRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class LiquorService {
  private final LiquorRepository liquorRepository;
  private final CategoryRepository categoryRepository;

  public Long save(SaveLiquorRequest request) {
    Category findCategory = categoryRepository.findById(request.getCategoryId()).orElseThrow(() -> new IllegalStateException("카테고리가 없습니다"));
    Liquor liquor = Liquor.builder().name(request.getName()).category(findCategory).price(request.getPrice()).rating(request.getRating()).build();
    liquorRepository.save(liquor);
    return liquor.getId();
  }

  public List<LiquorResponse> findAll() {
    return liquorRepository.findAllLiquor().stream().map((liquor) -> LiquorResponse
        .builder()
        .id(liquor.getId())
        .name(liquor.getName())
        .category(CategoryResponse.builder().id(liquor.getCategory().getId()).name(liquor.getCategory().getName()).build())
        .build()).toList();
  }

  public LiquorResponse findById(Long id) {
    Liquor findLiquor = liquorRepository.findById(id).orElseThrow(() -> new IllegalArgumentException("잘못된 아이디입니다."));
    return LiquorResponse
        .builder()
        .id(findLiquor.getId())
        .name(findLiquor.getName())
        .category(CategoryResponse.builder().id(findLiquor.getCategory().getId()).name(findLiquor.getName()).build())
        .build();
  }
}

File: src/main/kotlin/record/liquors/liquor/service/LiquorService.kt
package record.liquors.liquor.service

import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import record.liquors.liquor.api.LiquorResponse
import record.liquors.liquor.api.LiquorSaveRequest
import record.liquors.liquor.api.LiquorUpdateRequest
import record.liquors.liquor.entity.Liquor
import record.liquors.liquor.repository.LiquorCategoryRepository
import record.liquors.liquor.repository.LiquorRepository
import kotlin.NoSuchElementException
import kotlin.streams.toList

@Service
@Transactional(readOnly = true)
class LiquorService(
  val liquorRepository: LiquorRepository,
  val liquorCategoryRepository: LiquorCategoryRepository
) {
  @Transactional
  fun save(request: LiquorSaveRequest): Long {
    val liquor = Liquor(
      name = request.name,
      rating = request.rating,
      review = request.review,
      category = liquorCategoryRepository.findById(request.categoryId)
        .orElseThrow { throw NoSuchElementException("liquor category not found") }
    )
    liquorRepository.save(liquor)
    return liquor.id!!
  }

  fun findOne(id: Long): Liquor {
    return liquorRepository.findById(id).orElseThrow { throw NoSuchElementException("liquor not found") }
  }

  fun findLiquors(pageable: Pageable): Page<LiquorResponse> {
    return PageImpl(liquorRepository.findAll(pageable).stream().map { liquor -> LiquorResponse.toDto(liquor) }.toList())
  }

  @Transactional
  fun update(id: Long, request: LiquorUpdateRequest): Long {
    liquorRepository.findById(id).orElseThrow { throw NoSuchElementException("liquor not found") }.update(
      name = request.name,
      rating = request.rating,
      review = request.review,
      category = liquorCategoryRepository.findById(request.categoryId)
        .orElseThrow { throw NoSuchElementException("liquor category not found") }
    )
    return id
  }
}